src/Featurizers/Components/NormUpdaters.h

```diff
@@ -0,0 +1,413 @@
+};
+}
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::L1NormUpdater::L1NormResult
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Details::L1NormUpdater<T>::L1NormResult::L1NormResult(long double l1_norm) :
+    L1_norm(std::move(l1_norm)) {
+        // input type can only be integer or numeric
+        static_assert(Traits<T>::IsIntOrNumeric::value, "Input type to norm updater has to be integer or numerical types!");
+        if (L1_norm < 0) {
+            throw std::invalid_argument("l1 norm shouldn't be less than 0!");
+        }
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::L2NormUpdater::L2NormResult
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Details::L2NormUpdater<T>::L2NormResult::L2NormResult(long double l2_norm) :
+    L2_norm(std::move(l2_norm)) {
+        // input type can only be integer or numeric
+        static_assert(Traits<T>::IsIntOrNumeric::value, "Input type to norm updater has to be integer or numerical types!");
+        if (L2_norm < 0) {
+            throw std::invalid_argument("l2 norm shouldn't be less than 0!");
+        }
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::MaxNormUpdater::MaxNormResult
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Details::MaxNormUpdater<T>::MaxNormResult::MaxNormResult(typename TypeSelector::MaxNormTypeSelector<T>::type max_norm) :
+    Max_norm(std::move(max_norm)) {
+        // input type can only be integer or numeric
+        static_assert(Traits<T>::IsIntOrNumeric::value, "Input type to norm updater has to be integer or numerical types!");
+        if (Max_norm < 0) {
+            throw std::invalid_argument("max norm shouldn't be less than 0!");
+        }
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::L1NormUpdater
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Details::L1NormUpdater<T>::L1NormUpdater() :
+    _l1_norm(0),
+    _update_l1_flag(false) {
+        // input type can only be integer or numeric
+        static_assert(Traits<T>::IsIntOrNumeric::value, "Input type to norm updater has to be integer or numerical types!");

```

Comment:
static_asserts can appear anywhere in code, so it might be better to place these in the class declaration.